File: mobile/android/src/main/kotlin/com/otravelz/android/ui/screens/TransitViewModel.kt
package com.otravelz.android.ui.screens

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.otravelz.android.data.repository.TransitRepository
import com.otravelz.android.data.repository.TransitRepositoryImpl
import com.otravelz.android.domain.model.TransitRegion
import com.otravelz.android.domain.model.TransitRouteDetail
import com.otravelz.android.domain.model.TransitRouteSummary
import com.otravelz.android.domain.model.TransitSearchEngine
import com.otravelz.android.domain.model.TransitStop
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class TransitUiState(
    val isLoading: Boolean = true,
    val allRoutes: List<TransitRouteSummary> = emptyList(),
    val filteredRoutes: List<TransitRouteSummary> = emptyList(),
    val searchQuery: String = "",
    val selectedRegion: TransitRegion? = null,
    val selectedRouteDetail: TransitRouteDetail? = null,
    val isRouteDetailLoading: Boolean = false,
    val selectedDirectionIndex: Int = 0,
    val selectedStopForSheet: TransitStop? = null,
    val userLat: Double? = null,
    val userLon: Double? = null,
    val isRealGps: Boolean = false
) {
    val totalRouteCount: Int
        get() = allRoutes.size

    val filteredRouteCount: Int
        get() = filteredRoutes.size

    val hasActiveFilters: Boolean
        get() = searchQuery.isNotBlank() || selectedRegion != null
}

class TransitViewModel(
    application: Application,
    private val repository: TransitRepository = TransitRepositoryImpl(application)
) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(TransitUiState())
    val uiState: StateFlow<TransitUiState> = _uiState.asStateFlow()

    init {
        loadRoutes()
    }

    fun loadRoutes() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }
            val routes = repository.getRoutes()
            val filtered = TransitSearchEngine.filterAndRank(
                routes = routes,
                query = _uiState.value.searchQuery,
                selectedRegion = _uiState.value.selectedRegion
            )
            _uiState.update {
                it.copy(
                    isLoading = false,
                    allRoutes = routes,
                    filteredRoutes = filtered
                )
            }
        }
    }

    fun onSearchQueryChanged(query: String) {
        _uiState.update { current ->
            val filtered = TransitSearchEngine.filterAndRank(
                routes = current.allRoutes,
                query = query,
                selectedRegion = current.selectedRegion
            )
            current.copy(searchQuery = query, filteredRoutes = filtered)
        }
    }

    fun onRegionSelected(region: TransitRegion?) {
        _uiState.update { current ->
            val nextRegion = if (current.selectedRegion == region) null else region
            val filtered = TransitSearchEngine.filterAndRank(
                routes = current.allRoutes,
                query = current.searchQuery,
                selectedRegion = nextRegion
            )
            current.copy(selectedRegion = nextRegion, filteredRoutes = filtered)
        }
    }

    fun clearFilters() {
        _uiState.update { current ->
            val filtered = TransitSearchEngine.filterAndRank(
                routes = current.allRoutes,
                query = "",
                selectedRegion = null
            )
            current.copy(searchQuery = "", selectedRegion = null, filteredRoutes = filtered)
        }
    }

    fun selectRoute(routeId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isRouteDetailLoading = true, selectedDirectionIndex = 0) }
            val detail = repository.getRouteDetail(routeId)
            _uiState.update {
                it.copy(
                    isRouteDetailLoading = false,
                    selectedRouteDetail = detail
                )
            }
        }
    }

    fun clearSelectedRoute() {
        _uiState.update { it.copy(selectedRouteDetail = null, selectedStopForSheet = null) }
    }

    fun selectDirection(index: Int) {
        _uiState.update { it.copy(selectedDirectionIndex = index) }
    }

    fun selectStop(stop: TransitStop) {
        _uiState.update { it.copy(selectedStopForSheet = stop) }
    }

    fun dismissStopSheet() {
        _uiState.update { it.copy(selectedStopForSheet = null) }
    }

    fun updateUserLocation(lat: Double?, lon: Double?, isReal: Boolean) {
        _uiState.update { it.copy(userLat = lat, userLon = lon, isRealGps = isReal) }
    }
}

File: mobile/ios/OTravelz/Features/TransitViewModel.swift
import Foundation
import SwiftUI

@MainActor
public final class TransitViewModel: ObservableObject {
    @Published public var allRoutes: [TransitRouteSummary] = []
    @Published public var filteredRoutes: [TransitRouteSummary] = []
    @Published public var searchQuery: String = ""
    @Published public var selectedRegion: TransitRegion? = nil
    @Published public var isLoading: Bool = true
    @Published public var selectedRouteDetail: TransitRouteDetail? = nil
    @Published public var isRouteDetailLoading: Bool = false
    @Published public var selectedDirectionIndex: Int = 0
    @Published public var selectedStopForSheet: TransitStop? = nil
    @Published public var userLat: Double? = nil
    @Published public var userLon: Double? = nil
    @Published public var isRealGps: Bool = false

    private let repository: TransitRepositoryProtocol

    public init(repository: TransitRepositoryProtocol = TransitRepository.shared) {
        self.repository = repository
        loadRoutes()
    }

    public func loadRoutes() {
        isLoading = true
        Task {
            let routes = await repository.getRoutes()
            self.allRoutes = routes
            self.applyFilters()
            self.isLoading = false
        }
    }

    public func onSearchQueryChanged(_ query: String) {
        self.searchQuery = query
        applyFilters()
    }

    public func onRegionSelected(_ region: TransitRegion?) {
        if self.selectedRegion == region {
            self.selectedRegion = nil
        } else {
            self.selectedRegion = region
        }
        applyFilters()
    }

    public func clearFilters() {
        self.searchQuery = ""
        self.selectedRegion = nil
        applyFilters()
    }

    public func selectRoute(routeId: String) {
        isRouteDetailLoading = true
        selectedDirectionIndex = 0
        Task {
            let detail = await repository.getRouteDetail(routeId: routeId)
            self.selectedRouteDetail = detail
            self.isRouteDetailLoading = false
        }
    }

    public func clearSelectedRoute() {
        self.selectedRouteDetail = nil
        self.selectedStopForSheet = nil
    }

    public func selectDirection(index: Int) {
        self.selectedDirectionIndex = index
    }

    public func selectStop(_ stop: TransitStop) {
        self.selectedStopForSheet = stop
    }

    public func dismissStopSheet() {
        self.selectedStopForSheet = nil
    }

    public func updateUserLocation(lat: Double?, lon: Double?, isReal: Bool) {
        self.userLat = lat
        self.userLon = lon
        self.isRealGps = isReal
    }

    private func applyFilters() {
        self.filteredRoutes = TransitSearchEngine.filterAndRank(
            routes: allRoutes,
            query: searchQuery,
            selectedRegion: selectedRegion
        )
    }

    public var hasActiveFilters: Bool {
        !searchQuery.trimmingCharacters(in: .whitespaces).isEmpty || selectedRegion != nil
    }
}
